### PR TITLE
Fix manager.extend to revive support of non Extension object argument

### PR DESCRIPTION
**Pull request by asi1024**
https://github.com/pfnet/pytorch-pfn-extras/pull/200#discussion_r668716056

### CODE REVIEW

File: pytorch_pfn_extras/training/manager.py [hunk 1]
```diff
@@ -240,8 +240,7 @@ def extend(self, extension, name=None, trigger=None, priority=None,
 
         """
         if not isinstance(extension, extension_module.Extension):
-            raise TypeError(
-                f"'{type(extension)}' cannot be interpreted as an extension")
+            extension = extension_module.make_extension()(extension)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:243]: Could you add test for this?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Comment by asi1024** [comment]
@kmaehashi PTAL.

File: pytorch_pfn_extras/training/manager.py [hunk 2]
```diff
@@ -240,8 +240,15 @@ def extend(self, extension, name=None, trigger=None, priority=None,
...
         """
         if not isinstance(extension, extension_module.Extension):
-            raise TypeError(
-                f"'{type(extension)}' cannot be interpreted as an extension")
+            if getattr(extension, '__name__', None) != '<lambda>':
+                raise TypeError(
+                    "`extension` must be an Extension or a lambda function.\n"
+                    "1: Use `ppe.training.make_extension` which converts "
+                    "user-defined function to an Extension object, or\n"
+                    "2: Define custom extension class that overrides "
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:248]:
```suggestion
                    "2: Define custom extension class that inherits "
```

Comment (kmaehashi) [reply pytorch_pfn_extras/training/manager.py:248]: Could revert support for methods and functions as well?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/training/extension.py [hunk 3]
```diff
@@ -128,32 +128,27 @@ def load_state_dict(self, to_load):
 
 class _CallableExtension(Extension):
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:129]: I think now it's better to name it as:
```suggestion
class _WrappedExtension(Extension):
```

File: pytorch_pfn_extras/training/extension.py [hunk 4]
```diff
@@ -192,14 +187,16 @@ def make_extension(
...
+        ext.finalize = finalizer
+        ext.on_error = on_error
+        ext.initialize = initializer
+        return ext
 
     return decorator
+
+
+def _get_extension(ext) -> Extension:
+    return ext if isinstance(ext, Extension) else _CallableExtension(ext)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:202]:
```suggestion
def _as_extension(ext) -> Extension:
    return ext if isinstance(ext, Extension) else _WrappedExtension(ext)
```

File: pytorch_pfn_extras/training/manager.py [hunk 5]
```diff
@@ -239,9 +239,7 @@ def extend(self, extension, name=None, trigger=None, priority=None,
                 instead.
 
         """
-        if not isinstance(extension, extension_module.Extension):
-            raise TypeError(
-                f"'{type(extension)}' cannot be interpreted as an extension")
+        extension = extension_module._get_extension(extension)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:242]:
```suggestion
        extension = extension_module._as_extension(extension)
```

File: pytorch_pfn_extras/training/extension.py [hunk 6]
```diff
@@ -164,7 +159,7 @@ def make_extension(
         initializer=lambda manager: None,
         on_error=lambda manager, exc, tb: None,
 ):
-    """Decorator to make given functions into extensions.
+    """Decorator to make given functions into Extension object.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:162]: The decorator no longer returns Extension object.
```suggestion
    """Decorator to make given function into an extension.
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
LGTM except for minor comments!

**Review by kmaehashi — APPROVED** [review]

**Comment by kmaehashi** [comment]
Thanks, LGTM!

**Pull request merged**